src/states/node.rs

```diff
@@ -2388,36 +2393,46 @@ impl Node {
             self.stats.count_route(route);
         }
 
-        let routing_msg = signed_msg.routing_message();
+        let dst = signed_msg.routing_message().dst;
 
-        if let Authority::Client { ref peer_id, .. } = routing_msg.dst {
-            if *self.name() == routing_msg.dst.name() {
+        if let Authority::Client { ref peer_id, .. } = dst {
+            if *self.name() == dst.name() {
                 // This is a message for a client we are the proxy of. Relay it.
-                return self.relay_to_client(signed_msg.clone(), peer_id);
-            } else if self.in_authority(&routing_msg.dst) {
+                return self.relay_to_client(signed_msg, peer_id);
+            } else if self.in_authority(&dst) {
                 return Ok(()); // Message is for us as a client.
             }
         }
 
-        let (new_sent_to, target_peer_ids) = self.get_targets(routing_msg, route, hop, sent_to)?;
+        let (new_sent_to, target_peer_ids) =
+            self.get_targets(signed_msg.routing_message(), route, hop, sent_to)?;
 
         for target_peer_id in target_peer_ids {
-            self.send_signed_msg_to_peer(signed_msg, target_peer_id, route, new_sent_to.clone())?;
+            self.send_signed_msg_to_peer(signed_msg.clone(),
+                                         target_peer_id,
+                                         route,
+                                         new_sent_to.clone())?;
         }
         Ok(())
     }
 
+    // Filter, then convert the message to a `Hop` or `TunnelHop` `Message` and serialise.
+    // Send this byte string.
     fn send_signed_msg_to_peer(&mut self,
-                               signed_msg: &SignedMessage,
+                               signed_msg: SignedMessage,
                                target: PeerId,
                                route: u8,
                                sent_to: BTreeSet<XorName>)
                                -> Result<(), RoutingError> {
+        if self.filter_outgoing_routing_msg(signed_msg.routing_message(), &target, route) {
```

Comment:
If the message got filtered, then we already sent an identical version to this client, so were able to send to it then. I also don't think it's critical to immediately disconnect from a peer we don't have a route to (we ought to have noticed before if the route just dropped?).

I'm not 100% sure this would be equivalent in some edge-case, but the differences seem pretty unimportant.